### ISSUE TITLE
Determining file type

### ISSUE THREAD

**Issue by audrism** [ssc-oscar/clone-finder#1]
File type can be determined by extension but not always. 

Sometimes it is necessary to see what are 
- the file extensions of other files in the folder/repo
- the folder structure (e.g., presence of src usually indicates C/C++)
- the file content
- author (they may use file extensions in an idiosyncratic way)

Since many languages/technologie (file types) exist, instead of hand-crafing an algorithm for each type, it may be worth duing supervised learning: 
extract some features from each dimension listed above and train on a 
labeled data. 

Since file type is fairly important for clone detection, woud it make sense to try to do it well?

**Comment by KyLinXYL** [comment]
Yes, since we need to call the language specific segmentation engine, determining file type is an important part in this framework.

I think extension can determine most of file types and the segmentation engine must have the function of syntax analysis. So we can first determine the file type only by extension and use segmentation engine to check whether this determination is right. If the engine reports some errors, then we can use the machine learning algorithm to determine again.

Since we firmly believe that the dimension of extension is wrong, we may train this model without the dimension of extension but only with the other dimensions you have mentioned above.
